From f398c546c6fc43121131f41acec56b5a851bd35e Mon Sep 17 00:00:00 2001
From: Martin Wilck <mwilck@suse.com>
Date: Sat, 7 Apr 2018 17:33:48 +0200
Subject: [PATCH] systemd-udevd: limit children-max by available memory (#8668)

Udev workers consume typically 50-100MiB virtual memory.
On systems with lots of CPUs and relatively low memory, that may
easily cause workers to be OOM-killed.

This patch limits the number of workers to 8 per GiB memory.
But don't let the limit drop below the smallest value we had
without this patch (8 + 1 * 2 = 10); on small systems, udev's
memory footprint is likely lower.

(cherry picked from commit e438c57a640ac5afba366531be5e456b9fe22672)
---
 src/udev/udevd.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/udev/udevd.c b/src/udev/udevd.c
index 615c4ed3e2..daaab6417a 100644
--- a/src/udev/udevd.c
+++ b/src/udev/udevd.c
@@ -1676,12 +1676,16 @@ int main(int argc, char *argv[]) {
 
         if (arg_children_max == 0) {
                 cpu_set_t cpu_set;
+                unsigned long mem_limit;
 
                 arg_children_max = 8;
 
                 if (sched_getaffinity(0, sizeof(cpu_set), &cpu_set) == 0)
                         arg_children_max += CPU_COUNT(&cpu_set) * 2;
 
+                mem_limit = physical_memory() / (128LU*1024*1024);
+                arg_children_max = MAX(10U, MIN(arg_children_max, mem_limit));
+
                 log_debug("set children_max to %u", arg_children_max);
         }
 
-- 
2.17.0

